File: deps/src/upstream_examples/bench_assembly.cpp
#include <Tpetra_CrsMatrix.hpp>
#include <Tpetra_DefaultPlatform.hpp>
#include <Tpetra_Map.hpp>
#include <Tpetra_MultiVector.hpp>
#include <Tpetra_Vector.hpp>
#include <Tpetra_Version.hpp>
#include <Teuchos_Array.hpp>
#include <Teuchos_GlobalMPISession.hpp>
#include <Teuchos_oblackholestream.hpp>
#include <Teuchos_RCP.hpp>
#include <Teuchos_ScalarTraits.hpp>
#include <Teuchos_TimeMonitor.hpp>
#include <Teuchos_VerboseObject.hpp>


Teuchos::RCP<Teuchos::Time> graph_time;
Teuchos::RCP<Teuchos::Time> fill_time;

typedef double scalar_t;
typedef int local_t;
typedef int64_t global_t;


template<typename GraphT>
void build_graph(GraphT& A)
{
  Teuchos::TimeMonitor local_timer(*graph_time);
  const auto& rowmap = *A.getRowMap();
  local_t n_my_elms = rowmap.getNodeNumElements();
  const global_t gmin = rowmap.getMinGlobalIndex();
  const global_t gmax = rowmap.getMaxGlobalIndex();

  // storage for the per-row values
  global_t row_indices_arr[5] = {0,0,0,0,0};
  Teuchos::ArrayView<global_t> row_indices(row_indices_arr,5);

  for(local_t i = 0; i != n_my_elms; ++i)
  {
    const global_t global_row =  rowmap.getGlobalElement(i);
    for(int j = 0; j != 5; ++j)
    {
      row_indices[j] = global_row - (2-j);
      if(row_indices[j] < gmin || row_indices[j] > gmax)
      {
        row_indices[j] = global_row;
      }
    }
    A.insertGlobalIndices(global_row, row_indices);
  }
}

template<typename MatrixT>
void fill_matrix(MatrixT& A)
{
  Teuchos::TimeMonitor local_timer(*fill_time);
  const auto& rowmap = *A.getRowMap();
  const local_t n_my_elms = rowmap.getNodeNumElements();

  // storage for the per-row values
  local_t row_indices_arr[5] = {0,0,0,0,0};
  scalar_t row_values_arr[5] = {0,0,0,0,0};

  Teuchos::ArrayView<local_t> row_indices(row_indices_arr,5);
  Teuchos::ArrayView<scalar_t> row_values(row_values_arr,5);
  std::size_t n_elems;
  for(local_t row = 0; row != n_my_elms; ++row)
  {
    A.getLocalRowCopy(row, row_indices, row_values, n_elems);
    for (std::size_t j = 0; j != n_elems; ++j)
    {
      row_values[j] = static_cast<scalar_t>(j+1);
    }
    A.replaceLocalValues(row, row_indices, row_values);
  }
}


template<typename comm_t>
void bench_assembly(comm_t comm, const global_t nnodes)
{
  auto map = Teuchos::rcp(new Tpetra::Map<local_t,global_t>(nnodes, 0, comm));
  typedef Tpetra::Map<local_t,global_t>::node_type node_t;

  auto matrix_graph = Teuchos::rcp(new Tpetra::CrsGraph<local_t,global_t>(map, 5));
  build_graph(*matrix_graph);
  matrix_graph->fillComplete();

  // Matrix construction and fill
  auto A = Teuchos::rcp(new Tpetra::CrsMatrix<scalar_t,local_t,global_t>(matrix_graph));
  A->resumeFill();
  fill_matrix(*A);
  A->fillComplete();

  //A->describe(*Teuchos::VerboseObjectBase::getDefaultOStream(), Teuchos::VERB_EXTREME);
}

void do_run()
{
  graph_time = Teuchos::TimeMonitor::getNewCounter("Graph construction time");
  fill_time = Teuchos::TimeMonitor::getNewCounter("Matrix fill time");

  Teuchos::RCP<const Teuchos::Comm<int> > comm = Tpetra::DefaultPlatform::getDefaultPlatform ().getComm ();

  const global_t N = 1000000;
  
  for(int i = 0; i != 3; ++i)
  {
    bench_assembly(comm, N);
    Teuchos::TimeMonitor::report(std::cout);
    Teuchos::TimeMonitor::zeroOutTimers();
  }
}

int main (int argc, char *argv[])
{
  Teuchos::oblackholestream blackhole;
  Teuchos::GlobalMPISession mpiSession (&argc, &argv, &blackhole);
  
  do_run();

  return 0;
}

File: examples/bench_assembly.jl
using MPI
using Trilinos
using Test

function build_graph!(graph)
  rowmap = Tpetra.getRowMap(graph)
  n_my_elms = Tpetra.getNodeNumElements(rowmap)
  gmin = Tpetra.getMinGlobalIndex(rowmap)
  gmax = Tpetra.getMaxGlobalIndex(rowmap)

  row_indices = zeros(Int,5)
  row_indices_view = Teuchos.ArrayView(row_indices)

  for i in 0:n_my_elms-1
    global_row =  Tpetra.getGlobalElement(rowmap,i)
    for j in 1:5
      row_indices[j] = global_row - (3-j)
      if row_indices[j] < gmin || row_indices[j] > gmax
        row_indices[j] = global_row
      end
    end
      Tpetra.insertGlobalIndices(graph, global_row, row_indices)
  end
end

function fill_matrix!(A)
  rowmap = Tpetra.getRowMap(A)
  n_my_elms = Int(Tpetra.getNodeNumElements(rowmap))

  row_indices_arr = zeros(Int32,5)
  row_values_arr = zeros(5)

  row_indices = Teuchos.ArrayView(row_indices_arr)
  row_values = Teuchos.ArrayView(row_values_arr)
  n_elems = Ref(UInt(0))

  Tpetra.resumeFill(A)
  for row in 0:n_my_elms-1
    Tpetra.getLocalRowCopy(A, row, row_indices, row_values, n_elems)
    for j in 1:n_elems[]
      row_values_arr[j] = j
    end
    Tpetra.replaceLocalValues(A, row, row_indices, row_values)
  end
end

"""
Assemble the 2D Poisson problem on a structured grid
"""
function bench_assembly(comm, nnodes)
  # Construct map
  rowmap = Tpetra.Map(nnodes, 0, comm)

  # Construct graph (i.e. the sparsity pattern)
  matrix_graph = Tpetra.CrsGraph(rowmap, 5)
  println("-----------------------------------")
  println("Graph construction time:")
  @time build_graph!(matrix_graph)
  Tpetra.fillComplete(matrix_graph)

  # Matrix construction
  A = Tpetra.CrsMatrix(matrix_graph)
  Tpetra.resumeFill(A)
  println("Matrix fill time:")
  @time fill_matrix!(A)
  Tpetra.fillComplete(A)

  # This prints the matrix if uncommented
  #Tpetra.describe(A, Teuchos.VERB_EXTREME)
end

# MPI setup
if !MPI.Initialized()
  MPI.Init()
  MPI.finalize_atexit()
end
comm = Teuchos.MpiComm(MPI.CComm(MPI.COMM_WORLD))

my_rank = Teuchos.getRank(comm)
if my_rank != 0
  redirect_stdout(open("/dev/null", "w"))
  redirect_stderr(open("/dev/null", "w"))
end

N = 1000000
bench_assembly(comm, N)
bench_assembly(comm, N)
bench_assembly(comm, N)
